external-crates/move/move-compiler/src/sui_mode/typing.rs

```diff
@@ -114,18 +154,284 @@ fn module(
+                second_var.loc,
+                "'init' functions can have at most two parameters"
+            ),
+        ));
+    }
+}
+
+// While theoretically we could call this just once for the upper cased module struct, we break it
+// out into a separate function to help programmers understand the rules for one-time witness types,
+// when trying to write an 'init' function.
+fn check_otw_type(context: &mut Context, name: StructName, sdef: &N::StructDefinition) {
+    if context.one_time_witness.is_some() {
+        return;
+    }
+
+    let mut valid = true;
+    if let Some(tp) = sdef.type_parameters.first() {
+        let msg = "One-time witness types cannot have type parameters";
+        let mut diag = diag!(
+            OTW_DECL_DIAG,
+            (name.loc(), "Invalid one-time witness declaration"),
+            (tp.param.user_specified_name.loc, msg),
+        );
+        diag.add_note(OTW_NOTE);
+        context.env.add_diag(diag);
+        valid = false
+    }
+
+    if let N::StructFields::Defined(fields) = &sdef.fields {
+        if let Some(invalid_field_loc) = invalid_otw_field_loc(fields) {
+            let msg = format!(
+                "One-time witness types must have no fields, \
+                or exactly one field of type {}",
+                error_format(&Type_::bool(name.loc()), &Subst::empty())
+            );
+            let mut diag = diag!(
+                OTW_DECL_DIAG,
+                (name.loc(), "Invalid one-time witness declaration"),
+                (invalid_field_loc, msg),
+            );
+            diag.add_note(OTW_NOTE);
+            context.env.add_diag(diag);
+            valid = false
+        }
+    }
+
+    let invalid_ability_loc =
+        if !sdef.abilities.has_ability_(Ability_::Drop) || sdef.abilities.len() > 1 {
+            let loc = sdef
+                .abilities
+                .iter()
+                .find_map(|a| {
+                    if a.value != Ability_::Drop {
+                        Some(a.loc)
+                    } else {
+                        None
+                    }
+                })
+                .unwrap_or(name.loc());
+            Some(loc)
+        } else {
+            None
+        };
+    if let Some(loc) = invalid_ability_loc {
+        let msg = format!(
+            "One-time witness types can only have the have the '{}' ability",
+            Ability_::Drop
+        );
+        let mut diag = diag!(
+            OTW_DECL_DIAG,
+            (name.loc(), "Invalid one-time witness declaration"),
+            (loc, msg),
+        );
+        diag.add_note(OTW_NOTE);
+        context.env.add_diag(diag);
+        valid = false
+    }
+
+    context.one_time_witness = Some(if valid { Ok(name) } else { Err(()) })
+}
+
+// Find the first invalid field in a one-time witness type, if any.
+// First looks for a non-boolean field, otherwise looks for any field after the first.
+fn invalid_otw_field_loc(fields: &Fields<Type>) -> Option<Loc> {
```

Comment:
Nit. It took me a bit to parse it for correctness (which I think it is, and I am just slow...) but it probably wouldn't hurt adding a multi-field test.